host/sampi/http.go

```diff
@@ -0,0 +1,201 @@
+)
+
+// Cluster
+type Cluster struct {
+	state *State
+}
+
+func NewCluster(state *State) *Cluster {
+	return &Cluster{state}
+}
+
+// Scheduler Methods
+
+func (s *Cluster) ListHosts() ([]host.Host, error) {
+	hostMap := s.state.Get()
+	hostSlice := make([]host.Host, 0, len(hostMap))
+	for _, h := range hostMap {
+		hostSlice = append(hostSlice, h)
+	}
+	return hostSlice, nil
+}
+
+func (s *Cluster) AddJobs(req map[string][]*host.Job) (map[string]host.Host, error) {
+	s.state.Begin()
+	for host, jobs := range req {
+		if err := s.state.AddJobs(host, jobs); err != nil {
+			s.state.Rollback()
+			return nil, err
+		}
+	}
+	res := s.state.Commit()
+
+	for host, jobs := range req {
+		for _, job := range jobs {
+			s.state.SendJob(host, job)
+		}
+	}
+
+	return res, nil
+}
+
+// Host Service methods
+
+func (s *Cluster) RemoveJobs(hostID string, jobIDs ...string) error {
+	s.state.Begin()
+	s.state.RemoveJobs(hostID, jobIDs...)
+	s.state.Commit()
+	return nil
+}
+
+func (s *Cluster) StreamHostEvents(ch chan host.HostEvent, done chan bool) error {
+	s.state.AddListener(ch)
+	go func() {
+		<-done
+		go func() {
+			// drain to prevent deadlock while removing the listener
+			for range ch {
+			}
+		}()
+		s.state.RemoveListener(ch)
+		close(ch)
+	}()
+	return nil
+}
+
+type httpAPI struct {
+	cluster *Cluster
+}
+
+/* This will setup a SSEWriter + JSONEncoder and Write out the required Headers
+ * to the provided HTTP, the Encoder returned is used to write to the SSE
+ * Stream + JSON from data/objects/structs.
+ */
+func startJSONEventStreaming(w http.ResponseWriter) *json.Encoder {
+	wr := sse.NewSSEWriter(w)
+	enc := json.NewEncoder(httphelper.FlushWriter{Writer: wr, Enabled: true})
+	w.Header().Set("Content-Type", "text/event-stream; charset=utf-8")
+	w.WriteHeader(200)
+	wr.Flush()
+	return enc
+}
+
+// HTTP Route Handles
+func (c *httpAPI) ListHosts(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+	ret, err := c.cluster.ListHosts()
+	if err != nil {
+		httphelper.Error(w, err)
+		return
+	}
+	httphelper.JSON(w, 200, ret)
+}
+
+func (c *httpAPI) RegisterHost(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	h := &host.Host{}
+	if err := json.NewDecoder(r.Body).Decode(&h); err != nil {
+		httphelper.Error(w, err)
+		return
+	}
+
+	if h.ID == "" {
+		httphelper.Error(w, errors.New("sampi: host id must not be blank"))
+		return
+	}
+
+	ch := make(chan *host.Job)
+
+	c.cluster.state.Begin()
+	if c.cluster.state.HostExists(h.ID) {
+		c.cluster.state.Rollback()
+		httphelper.Error(w, errors.New("sampi: host exists"))
+		return
+	}
+	c.cluster.state.AddHost(h, ch)
+	c.cluster.state.Commit()
+	go c.cluster.state.sendEvent(h.ID, "add")
+
+	// "defer" cleanups in a goroutine that waits until the http stream is closed.
+	go func() {
+		<-w.(http.CloseNotifier).CloseNotify()
+		c.cluster.state.Begin()
+		c.cluster.state.RemoveHost(h.ID)
+		c.cluster.state.Commit()
+		go c.cluster.state.sendEvent(h.ID, "remove")
```

Comment:
Any reason to run this in a goroutine? We're already detached from the handler goroutine.